### PR TITLE
out_stackdriver: support gzip compression

### PR DESCRIPTION
**Pull request by CatherineF-dev**
----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:
- [x] Example configuration file for the change
- [x] Debug log output from testing the change

- [ ] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

If this is a change to packaging of containers or native binaries then please confirm it works for all targets.
- [ ] Run [local packaging test](./packaging/local-build-all.sh) showing all targets (including any new ones) build.
- [ ] Set `ok-package-test` label to test for all targets (requires maintainer to do).

**Documentation**

- [x] Documentation required for this feature

**Backporting**

- [ ] Backport to latest stable release.

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

### CODE REVIEW

**Comment by CatherineF-dev** [comment]
- Configuration example:
```
     [OUTPUT]
         # https://docs.fluentbit.io/manual/pipeline/outputs/stackdriver
         Name                      stackdriver
         compress_gzip             1
```

- Debug output:
Compression rate is about 95%, from 44562525 bytes to \2141049 bytes.

**Comment by edsiper** [comment]
quick review:

- rename the new configuration option to `compress` and support a string value. All output plugins that supports compression uses the same configuration property name, so we don't name it differently, as an example take a look at the Splunk plugin: https://github.com/fluent/fluent-bit/blob/master/plugins/out_splunk/splunk.c#L734-L738 

- sign-off your commits (DCO error)

once those are fixed we can do another review

**Comment by CatherineF-dev** [comment]
cc @qingling128

**Comment by CatherineF-dev** [comment]
It seems splunk is using `ctx->compress_gzip`.

https://github.com/fluent/fluent-bit/blob/master/plugins/out_splunk/splunk.c#L600

**Comment by CatherineF-dev** [comment]
Hi team, have changed to `compress`

**Comment by patrick-stephens** [comment]
Can you link the docs PR?

**Comment by CatherineF-dev** [comment]
### Context
#### Objective
Reduce network traffic when sending logs from k8s cluster into Cloud Logging.
It reduced bytes of logs by 95% after using gzip, from 100 bytes  to 5 bytes.
#### Background
Fluentbit stackriver_output is using `Content-Type: application/json` without compression.

### Design
#### Stackdriver output plugin API:
```
[OUTPUT]
         #[ https://docs.fluentbit.io/manual/pipeline/outputs/stackdriver](https://docs.fluentbit.io/manual/pipeline/outputs/stackdriver)
         Name                      stackdriver
         ...
         compress                  "gzip"
```

#### The HTTP header
```
Content-Type: application/json
Content-Encoding: gzip
```

cc @patrick-stephens This is the design doc.

**Comment by patrick-stephens** [comment]
@CatherineF-dev apologies, I mean the PR to provide the update to the Fluent Bit documentation. Typically we link the two together so when this PR is merged then released we can merge and release the docs for people using it.

You have added a new configuration parameter `compress` so this needs to be documented here:  https://docs.fluentbit.io/manual/pipeline/outputs/stackdriver#configuration-parameters

To do that, update the Markdown here and submit a PR: https://github.com/fluent/fluent-bit-docs/blob/master/pipeline/outputs/stackdriver.md

**Comment by CatherineF-dev** [comment]
@patrick-stephens Done, thx!

https://github.com/fluent/fluent-bit-docs/pull/1097

File: plugins/out_stackdriver/stackdriver.c
Comment (leonardo-albertovich) [inline plugins/out_stackdriver/stackdriver.c:2524]: Please revert the changes related to `payload_buf`, add a third variable to hold the final payload (and one for its sice) that's conditionally set to either a copy the raw result of `stackdriver_format` or the output buffer `flb_gzip_compress` produces to make the usage transparent while retaining a proper cleanup procedure.

File: plugins/out_stackdriver/stackdriver.c
Comment (leonardo-albertovich) [inline plugins/out_stackdriver/stackdriver.c:2509]: [flb_gzip_compress](https://github.com/fluent/fluent-bit/blob/ba49c5baa998f5187c6224ebcaf400de9f98370a/src/flb_gzip.c#L85) returns a buffer allocated with `flb_malloc` which is not compatible with `flb_sds_destroy`

File: plugins/out_stackdriver/stackdriver.c
Comment (leonardo-albertovich) [inline plugins/out_stackdriver/stackdriver.c:2622]: This is related to what I previously mentioned and to make things clearer it'd be better if you renamed the variable to something like `compressed_payload_buffer` which would clearly convey its purpose.

**Review by leonardo-albertovich — CHANGES_REQUESTED** [review]

**Comment by leonardo-albertovich** [comment]
The splunk used `compress_gzip` as the field name in the context structure so it only compares the string once when the plugin is initialized. That's somewhat understandable under the presumption that other algorithms won't ever be used, however, a better approach would've been defining a set of constants for different supported  algorithms (even if  that meant just defining one option for gzip and one for none), defining an integer field to store the interpreted value and using that in the flush function.

I think this will probably be streamlined in the near future. At the moment we only support gzip and snappy but we will add more algorithms soon (zstd being the first one that comes to mind).

**Comment by leonardo-albertovich** [comment]
Ideally, when we do that we should all of the code related to compression to `flb_http_client` so it's transparent to the plugin and we break this code duplication cycle.

**Comment by JeffLuoo** [comment]
Hi @leonardo-albertovich, what's the plan for decoupling the compression logic to the fluent-bit client? If that's not near future, I think we still need to have the compression in the out_stackdriver for now.

File: plugins/out_stackdriver/stackdriver.c
Comment (CatherineF-dev) [reply plugins/out_stackdriver/stackdriver.c:2509]: Thanks! Done.

File: plugins/out_stackdriver/stackdriver.c
Comment (CatherineF-dev) [reply plugins/out_stackdriver/stackdriver.c]: Thanks! Done.

**Review by leonardo-albertovich — DISMISSED** [review]

**Comment by CatherineF-dev** [comment]
/hold

Plan to add a safe check for NULL compressed
```
    if (compressed == FLB_TRUE) {
        flb_free(compressed_payload_buffer);
    }
```

**Comment by CatherineF-dev** [comment]
@leonardo-albertovich Hi, could you approve again? Thanks!

I rebased and added a safe check for `flb_free(compressed_payload_buffer)`.

Because tests failed before, try rebasing to see whether these failures will be fixed or not.

**Review by leonardo-albertovich — DISMISSED** [review]

**Comment by CatherineF-dev** [comment]
Hi @JeffLuoo, have you encountered similar test failures before?

I haven't found useful error information.

**Comment by CatherineF-dev** [comment]
Hi team, how to merge this PR?

I saw other PRs are merge manually.

**Comment by CatherineF-dev** [comment]
cc @qingling128 Need a LGTM on gzip for out_stackdriver plugin, thanks!

**Comment by leonardo-albertovich** [comment]
I have already approved this PR, @edsiper will merge it when he's able to.

**Comment by CatherineF-dev** [comment]
Ping ~

**Comment by igorpeshansky** [comment]
> Plan to add a safe check for NULL compressed
> 
> ```
>     if (compressed == FLB_TRUE) {
>         flb_free(compressed_payload_buffer);
>     }
> ```

BTW, `flb_free` uses `free` [1], which should already handle `NULL` properly [2].

[1] https://github.com/fluent/fluent-bit/blob/master/include/fluent-bit/flb_mem.h#L127 \
[2] https://linux.die.net/man/3/free

File: plugins/out_stackdriver/stackdriver.h
Comment (CatherineF-dev) [reply plugins/out_stackdriver/stackdriver.h:165]: Later refactoring.

File: plugins/out_stackdriver/stackdriver_conf.c
Comment (CatherineF-dev) [reply plugins/out_stackdriver/stackdriver_conf.c:556]: Good point. Updated to `FLB_FALSE`. 
Thx!

File: plugins/out_stackdriver/stackdriver_conf.c
Comment (CatherineF-dev) [reply plugins/out_stackdriver/stackdriver_conf.c:560]: Got it, thx!

File: plugins/out_stackdriver/stackdriver.c
Comment (CatherineF-dev) [reply plugins/out_stackdriver/stackdriver.c:2533]: Done.

File: plugins/out_stackdriver/stackdriver.c
Comment (CatherineF-dev) [reply plugins/out_stackdriver/stackdriver.c:2541]: A great point. Done.

File: plugins/out_stackdriver/stackdriver.c
Comment (CatherineF-dev) [reply plugins/out_stackdriver/stackdriver.c:2527]: Good point. Updated to `compressed_payload_buffer = payload_buf;`

**Comment by CatherineF-dev** [comment]
Thanks, have updated! @igorpeshansky

Okay, we can remove this safe check for `flb_free` in future PR.
https://github.com/fluent/fluent-bit/blob/6621b4bbeb2005f928e7ae87114bfe8bdc505def/plugins/out_stackdriver/stackdriver.c#L62-L64

File: plugins/out_stackdriver/stackdriver_conf.c
Comment (igorpeshansky) [inline plugins/out_stackdriver/stackdriver_conf.c]: Couldn't find the "Unresolve conversation" button, so adding this here to flag that there was a follow-up comment in the earlier (resolved) thread.

File: plugins/out_stackdriver/stackdriver.c
Comment (igorpeshansky) [reply plugins/out_stackdriver/stackdriver.c:2572]: The `Content-Type: text/plain` on that page is just an example. In this case, the content type is `application/json`, so that's what you should set it to (with `Content-Encoding: gzip`). Setting it to `text/plain` is wrong. Let's instead figure out why `application/json` didn't work.

File: plugins/out_stackdriver/stackdriver.c
Comment (igorpeshansky) [inline plugins/out_stackdriver/stackdriver.c:2768]: I think your indentation is off…
```suggestion
      FLB_CONFIG_MAP_STR, "compress", NULL,
      0, FLB_FALSE, 0,
```

File: plugins/out_stackdriver/stackdriver.h
Comment (igorpeshansky) [inline plugins/out_stackdriver/stackdriver.h:163]: You can now remove this field…
```suggestion
```

File: plugins/out_stackdriver/stackdriver.c
Comment (CatherineF-dev) [reply plugins/out_stackdriver/stackdriver.c:2572]: If it's wrong, what will happen? It won't be json in cloud logging?

I can check cloud logging UI.

Comment (CatherineF-dev) [reply plugins/out_stackdriver/stackdriver.c]: Verified, json is working

**Comment by igorpeshansky** [comment]
cc @leonardo-albertovich as well.

**Pull request merged**
